flink-1.17-connector-bigquery/flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/table/BigQueryDynamicTableSink.java

```diff
@@ -0,0 +1,86 @@
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package com.google.cloud.flink.bigquery.table;
+
+import org.apache.flink.annotation.Internal;
+import org.apache.flink.annotation.VisibleForTesting;
+import org.apache.flink.table.connector.ChangelogMode;
+import org.apache.flink.table.connector.sink.DynamicTableSink;
+import org.apache.flink.table.connector.sink.SinkV2Provider;
+import org.apache.flink.table.types.logical.LogicalType;
+
+import com.google.cloud.flink.bigquery.sink.BigQuerySink;
+import com.google.cloud.flink.bigquery.sink.BigQuerySinkConfig;
+import com.google.cloud.flink.bigquery.sink.serializer.BigQuerySchemaProviderImpl;
+import com.google.cloud.flink.bigquery.sink.serializer.BigQueryTableSchemaProvider;
+import com.google.cloud.flink.bigquery.sink.serializer.RowDataToProtoSerializer;
+import org.apache.avro.Schema;
+
+/** A {@link org.apache.flink.table.connector.sink.DynamicTableSink} for Google BigQuery. Tho */
+@Internal
+public class BigQueryDynamicTableSink implements DynamicTableSink {
+
+    private final BigQuerySinkConfig sinkConfig;
+    private final LogicalType logicalType;
+
+    public BigQueryDynamicTableSink(BigQuerySinkConfig sinkConfig, LogicalType logicalType) {
+        this.logicalType = logicalType;
+        Schema avroSchema =
+                BigQueryTableSchemaProvider.getAvroSchemaFromLogicalSchema(this.logicalType);
+        this.sinkConfig =
+                BigQuerySinkConfig.newBuilder()
+                        .schemaProvider(new BigQuerySchemaProviderImpl(avroSchema))
+                        .deliveryGuarantee(sinkConfig.getDeliveryGuarantee())
+                        .connectOptions(sinkConfig.getConnectOptions())
+                        .serializer(sinkConfig.getSerializer())
+                        .build();
+    }
+
+    @Override
+    public ChangelogMode getChangelogMode(ChangelogMode requestedMode) {
+        return ChangelogMode.insertOnly();
+    }
+
+    @Override
+    public SinkRuntimeProvider getSinkRuntimeProvider(Context context) {
+        // init() should be called itself.
+        // Set the logical type.
+        ((RowDataToProtoSerializer) sinkConfig.getSerializer()).setLogicalType(this.logicalType);
+        // Get the Datastream-API Sink.
+        return SinkV2Provider.of(BigQuerySink.get(this.sinkConfig, null));
+    }
+
+    @Override
+    public DynamicTableSink copy() {
+        return new BigQueryDynamicTableSink(this.sinkConfig, this.logicalType);
+    }
+
+    @Override
+    public String asSummaryString() {
+        return "BigQuery";
```

Comment:
At some point we should consider moving these hardcoded strings into a resource bundle file.

Reply:
Added, pls verify.